src/PermMatrix.jl

```diff
@@ -42,26 +43,74 @@ struct PermMatrix{Tv,Ti<:Integer,Vv<:AbstractVector{Tv},Vi<:AbstractVector{Ti}}
         new{Tv,Ti,Vv,Vi}(perm, vals)
     end
 end
-
-function PermMatrix{Tv,Ti}(perm, vals) where {Tv,Ti<:Integer}
-    PermMatrix{Tv,Ti,Vector{Tv},Vector{Ti}}(Vector{Ti}(perm), Vector{Tv}(vals))
+basetype(pm::PermMatrix) = PermMatrix
+Base.getindex(M::PermMatrix{Tv}, i::Integer, j::Integer) where {Tv} =
+    M.perm[i] == j ? M.vals[i] : zero(Tv)
+@propagate_inbounds function Base.setindex!(M::PermMatrix, val, i::Integer, j::Integer)
+    @assert M.perm[i] == j "Can not set index due to the absense of entry: ($i, $j)"
+    @inbounds M.vals[i] = val
 end
 
-function PermMatrix(
-    perm::Vi,
-    vals::Vv,
-) where {Tv,Ti<:Integer,Vv<:AbstractVector{Tv},Vi<:AbstractVector{Ti}}
-    PermMatrix{Tv,Ti,Vv,Vi}(perm, vals)
+# the column major version of `PermMatrix`
+struct PermMatrixCSC{Tv,Ti<:Integer,Vv<:AbstractVector{Tv},Vi<:AbstractVector{Ti}} <:
+       AbstractPermMatrix{Tv,Ti}
+    perm::Vi   # new orders
+    vals::Vv   # multiplied values.
+
+    function PermMatrixCSC{Tv,Ti,Vv,Vi}(
+        perm::Vi,
+        vals::Vv,
+    ) where {Tv,Ti<:Integer,Vv<:AbstractVector{Tv},Vi<:AbstractVector{Ti}}
+        if length(perm) != length(vals)
+            throw(
+                DimensionMismatch(
+                    "permutation ($(length(perm))) and multiply ($(length(vals))) length mismatch.",
+                ),
+            )
+        end
+        new{Tv,Ti,Vv,Vi}(perm, vals)
+    end
+end
+basetype(pm::PermMatrixCSC) = PermMatrixCSC
+@propagate_inbounds function Base.getindex(M::PermMatrixCSC{Tv}, i::Integer, j::Integer) where {Tv}
```

Comment:
this is correct, but others are wrong..., you need to pair up `propgate_inbounds` with `boundscheck` so that `boundscheck` can be cancelled inside a loop using `@inbounds`

Reply:
I see you point. Thanks! I have removed unnessesary `propagate_inbounds`.